### PR TITLE
Source API

### PR DESCRIPTION
**Pull request by rodjek**
Opening this to discuss how I think the Source API should look.

What I really want to prevent is sources turning into implementation clusterfucks ala librarian.  I'd like the Sources to only define a handful of required public methods:
- `fetch!`
- `versions`
- `dependencies` (maybe?  if we only fetch dependencies from the metadata files in the fetched modules we can possibly even abstract this away into `Henson::Source::Generic`

Then in `Henson::Source::Generic` we have the meat of the process
- `fetch_location` - If caching enabled, returns the cache location for the module, otherwise a temporary path
- `install!` - Install the module from `fetch_location` to the final destination
- etc

Basically, I want the Source objects to be as simple as possible - grab data and put it into a standard location (module contents -> standard path on disk, metadata -> instance vars).

### CODE REVIEW

File: lib/henson.rb [hunk 1]
```diff
@@ -28,4 +28,6 @@ def self.exit_code(i)
 
   class PuppetfileError    < Error; exit_code(14); end
   class PuppetfileNotFound < Error; exit_code(16); end
+  class ModuleNotFound     < Error; exit_code(18); end
+  class RequirementNotSatisfied < Error; exit_code(20); end
```
Comment (wfarr) [inline lib/henson.rb:36]: Alignment, because I'm picky.

File: lib/henson/dsl.rb [hunk 2]
```diff
@@ -26,10 +26,13 @@ def evaluate(puppetfile)
 
     def mod(name, version, opts = {})
       PuppetModule.new(name, version, opts).tap do |puppet_module|
-        puppet_module.fetch! unless puppet_module.source.kind_of? Source::Path
+        unless puppet_module.satisfied?
+          raise RequirementNotSatisfied, "#{name} (#{version})"
+        end
+        puppet_module.source.fetch!
```
Comment (wfarr) [inline lib/henson/dsl.rb]: I really feel that `puppet_module` ought to own fetch here, which can delegate to the source's fetch because Law of Demeter and all that.

File: lib/henson/source/path.rb [hunk 3]
```diff
@@ -14,6 +24,25 @@ def valid?
       def path_exists?
         path && File.directory?(path)
       end
+
+      def version_from_modulefile
```
Comment (wfarr) [inline lib/henson/source/path.rb:44]: I'm kind of contemplating adding multiple DSL classes, one of which would be a `DSL::Modulefile`. In this case, we could just define `version` in that DSL and then set `method_missing` there to just silently ignore any other bits.

File: spec/henson/source/path_spec.rb [hunk 4]
```diff
@@ -16,24 +16,31 @@
   context "valid?" do
     it "returns true if path_exists? is true" do
       instance.stubs(:path_exists?).returns(true)
-      instance.valid?.should be_true
+      instance.send(:valid?).should be_true
```
Comment (wfarr) [inline spec/henson/source/path_spec.rb]: Why all the `send`s here?

**Comment by wfarr** [comment]
:+1: on light sources — totally my intent also. A Source should only know:
- what versions are available
- if the version requested is currently installed
- how to install a version
- how to cache an installed version

---

Dependency lookup should be managed by `Source::Generic`. We should have a cascade of preferred dependency places we look at:
1. a `Modulefile`
2. a `Puppetfile`

These will both need to leverage DSL classes specific to their file.

**Comment by rodjek** [comment]
> Dependency lookup should be managed by `Source::Generic`

Totally agree.  I sketched out that method in `Source::Path` but it felt wrong to have it there.

File: spec/henson/source/path_spec.rb (hunk 4 above)
Comment (rodjek) [inline spec/henson/source/path_spec.rb]: Changed them to be private methods to make the public interface of the sources cleaner

File: lib/henson/source/path.rb (hunk 3 above)
Comment (rodjek) [inline lib/henson/source/path.rb:44]: I'm down with that, this was just a quick hack to show my intentions.

File: lib/henson/dsl.rb (hunk 2 above)
Comment (rodjek) [inline lib/henson/dsl.rb]: Agreed

File: lib/henson/source/path.rb (hunk 3 above)
Comment (wfarr) [inline lib/henson/source/path.rb:44]: Rad, can always refactor later. Please drop a `TODO` comment to that effect.

File: spec/henson/source/path_spec.rb (hunk 4 above)
Comment (wfarr) [inline spec/henson/source/path_spec.rb]: Dig it. There's other places in the code that could likely use that treatment as well.

**Comment by wfarr** [comment]
Rad.

**Comment by wfarr** [comment]
Merge/rebase master to get the simple delegator lulz.

**Comment by wfarr** [comment]
This is now in-line with master and I've (obviously) refactored a bunch of stuff.

I really want to knuckle down and "finish" this PR and merge it into master.

Here's what I think is remaining/want to get done to merge this:
- Eliminate `Source::Path#version_from_modulefile` and instead use the DSL inferred version
- Make dependencies on `DSL::Modulefile` into `PuppetModule`s instead of hashes.
- Make `Installer` check for Modulefile first, then Puppetfile, and otherwise raise an error about no metadata source.

From there we can do smaller PRs for remaining things on master, otherwise this PR will just go epic and be confusing.

File: lib/henson/installer.rb [hunk 5]
```diff
@@ -1,9 +1,22 @@
...
+require "fileutils"
 
 module Henson
   class Installer
     def self.install!
-      parse_puppetfile!
+      FileUtils.mkdir_p File.expand_path(Henson.settings[:path])
+
+      parse_puppetfile!.modules.each do |mod|
+        mod.fetch! if mod.needs_fetching?
```
Comment (rodjek) [inline lib/henson/installer.rb:10]: We'll probably need to split this out into two loops, one to fetch and then one to install as we'll eventually have a dependency resolution phase that'll have to happen after fetch but before install.

Comment (wfarr) [inline lib/henson/installer.rb:10]: Probably right.

**Comment by rodjek** [comment]
Looks good, I say we merge this behemoth :)

**Comment by wfarr** [comment]
Merging.

Then let's focus on getting the path source support with transitive dependencies working (assuming path sources only on both sides), and then implementing the same sauce for the git source.
